worker/upload.go

```diff
@@ -187,6 +229,166 @@ func (w *worker) initUploadManager(maxOverdrive uint64, overdriveTimeout time.Du
 	w.uploadManager = newUploadManager(w.bus, w, w, maxOverdrive, overdriveTimeout, logger)
 }
 
+func (w *worker) upload(ctx context.Context, r io.Reader, bucket, path string, opts ...UploadOption) (string, error) {
+	//  build upload parameters
+	up := defaultParameters()
+	for _, opt := range opts {
+		opt(&up)
+	}
+
+	// perform the upload
+	obj, partialSlabData, used, etag, err := w.uploadManager.Upload(ctx, r, up)
+	if err != nil {
+		return "", fmt.Errorf("couldn't upload object: %w", err)
+	}
+
+	// add parital slabs
+	if len(partialSlabData) > 0 {
+		partialSlabs, err := w.bus.AddPartialSlab(ctx, partialSlabData, uint8(up.rs.MinShards), uint8(up.rs.TotalShards), up.contractSet)
+		if err != nil {
+			return "", err
+		}
+		obj.PartialSlabs = partialSlabs
+	}
+
+	// persist the object
+	err = w.bus.AddObject(ctx, bucket, path, up.contractSet, obj, used)
+	if err != nil {
+		return "", fmt.Errorf("couldn't add object: %w", err)
+	}
+
+	// if packing was enabled try uploading packed slabs in a separate goroutine
+	if up.packing {
+		go w.tryUploadPackedSlabs(up.rs, up.contractSet)
+	}
+	return etag, nil
+}
+
+func (w *worker) uploadMultiPart(ctx context.Context, r io.Reader, bucket, path, uploadID string, partNumber int, opts ...UploadOption) (string, error) {
+	//  build upload parameters
+	up := defaultParameters()
+	for _, opt := range opts {
+		opt(&up)
+	}
+
+	// upload the part
+	obj, partialSlabData, used, etag, err := w.uploadManager.Upload(ctx, r, up)
+	if err != nil {
+		return "", fmt.Errorf("couldn't upload object: %w", err)
+	}
+
+	// add parital slabs
+	if len(partialSlabData) > 0 {
+		partialSlabs, err := w.bus.AddPartialSlab(ctx, partialSlabData, uint8(up.rs.MinShards), uint8(up.rs.TotalShards), up.contractSet)
+		if err != nil {
+			return "", err
+		}
+		obj.PartialSlabs = partialSlabs
+	}
+
+	// persist the part
+	err = w.bus.AddMultipartPart(ctx, bucket, path, up.contractSet, uploadID, partNumber, obj.Slabs, obj.PartialSlabs, etag, used)
+	if err != nil {
+		return "", fmt.Errorf("couldn't add multi part: %w", err)
+	}
+
+	// if packing was enabled try uploading packed slabs in a separate goroutine
+	if up.packing {
+		go w.tryUploadPackedSlabs(up.rs, up.contractSet)
+	}
+	return etag, nil
+}
+
+func (w *worker) tryUploadPackedSlabs(rs api.RedundancySettings, contractSet string) {
+	key := fmt.Sprintf("%d-%d_%s", rs.MinShards, rs.TotalShards, contractSet)
+
+	w.uploadsMu.Lock()
+	if w.uploadingPackedSlabs[key] {
+		w.uploadsMu.Unlock()
+		return
+	}
+	w.uploadingPackedSlabs[key] = true
+	w.uploadsMu.Unlock()
+
+	// make sure we mark uploading packed slabs as false when we're done
+	defer func() {
+		w.uploadsMu.Lock()
+		w.uploadingPackedSlabs[key] = false
+		w.uploadsMu.Unlock()
+	}()
+
+	// keep uploading packed slabs until we're done
+	packedSlabsFound := true
+	for packedSlabsFound {
+		packedSlabsFound = func() bool {
```

Comment:
Same as below. This loop is very long and has 2 inline functions. Would be nice to clean that up a bit.